File: Sources/XYColorOC/Core/UIView+PrivateView.m
//
//  UIView+PrivateView.m
//  XYColor
//
//  Created by RayJiang on 2019/11/20.
//  Copyright © 2019 RayJiang. All rights reserved.
//

#import "UIView+PrivateView.h"
#import "_XYColor_PrivateView.h"
#import <objc/runtime.h>

@implementation UIView (PrivateView)

static const char PrivateViewKey = '\0';

- (void)setPv:(_XYColor_PrivateView *)pv {
    if (pv != self.pv) {
        [self.pv removeFromSuperview];
        pv.hidden = YES;
        [self insertSubview:pv atIndex:0];
        objc_setAssociatedObject(self, &PrivateViewKey, pv, OBJC_ASSOCIATION_RETAIN);
    }
}

- (_XYColor_PrivateView *)pv {
    return objc_getAssociatedObject(self, &PrivateViewKey);
}

@end

File: Sources/XYColor/Core/UIView+PrivateView.swift
//
//  UIView+PrivateView.swift
//  XYColor
//
//  Created by RayJiang on 2019/11/20.
//  Copyright © 2019 RayJiang. All rights reserved.
//

import UIKit

private var _privateViewKey: UInt8 = 0

extension UIView {
    
    var _pv: _PrivateView? {
        get {
            return objc_getAssociatedObject(self, &_privateViewKey) as? _PrivateView
        }
        set {
            if let newValue = newValue, _pv != newValue {
                _pv?.removeFromSuperview()
                newValue.isHidden = true
                insertSubview(newValue, at: 0)
                objc_setAssociatedObject(self, &_privateViewKey, newValue, .OBJC_ASSOCIATION_ASSIGN)
            }
        }
    }
}
